---
title: "Final Project, RNA-seq: DE Analysis of AKAP11 Mutant Mice"
output:
  pdf_document: default
  html_document: default
---
### Download the neccessary packages:
```{r package-download}
BiocManager::install("fgsea")
```
### Load the libraries:
```{r library-load}
library(DESeq2)
library(tidyverse)
library(ggplot2)
library(dplyr)
library(fgsea)
library(msigdbr)
```

## DE Analysis

### Reading in the counts matrix

Conversion to a readable matrix. 
```{r counts}
cts <- as.matrix(read.csv("counts_matrix.csv", row.names='gene', check.names = FALSE)) #check nqmes since sample names start with numbers
nrow(cts)
head(cts, 10)
```

### Constructing the colData

Here there are three factors of interest: KO (-/-), HET(-/+) and WT (control) samples.

```{r coldata}
coldata <- read.csv("colData.csv") #self-made csv to put samples in order as shown in counts matrix
coldata
```

### Design Formula

Our design formula will be ~ genotype.

## Filtering the counts matrix

Here we will filter the counts by keeping genes with at least 10 counts in at least 3 samples, a light pre-filtering strategy with the intention of removing noise from our matrix. After filtering, we go from 78334 genes to 22563.      

```{r filtering}
keep <- rowSums(cts >= 10) >= 3
counts_filtered <- cts[keep, ]
num_filtered <- nrow(counts_filtered)
num_filtered



```

## Constructing the DESeq object

We use the provided DESeq2 library to harness the function "DESeqDataSetFromMatrix" in order to create a DESeq object. This is later placed into the DESeq function. The three elements required to construct the object: the counts, the coldata, and the design. A reminder that there are 3 levels in our case. 

```{r deseq-object}

coldata$genotype <- factor(coldata$genotype,
                           levels = c("WT", "HET", "KO"))

dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              colData = coldata,
                              design = ~ genotype)
```

## Running DESeq

We run DESeq2 on the created object and produce results based upon our design. Our results are based on three comparisons: KO to HET, HET to WT, and KO to WT. 
```{r dds-run, echo = FALSE}
dds <- DESeq(dds)

#define contrasts
res_KO_vs_WT <- results(dds, contrast = c("genotype", "KO", "WT"))

res_HET_vs_WT <- results(dds, contrast = c("genotype", "HET", "WT"))

res_KO_vs_HET <- results(dds, contrast = c("genotype", "KO", "HET"))
```

## Generating QC Plots: Mean-Average (MA) and PCA

### MA Plot

Here we use MA plots to visualize RNA-seq technical quality and get a general overview of expression:
```{r plot-MA_KO_vs_WT}
png("MAplot_KOvsWT.png", width = 8, height = 6, units = "in", res = 300)
MA_KOvsWT <- plotMA(res_KO_vs_WT, ylim = c(-3, 3), main = "MA Plot: KO vs WT")
dev.off()
```

```{r plot-MA_HET_vs_WT}
png("MAplot_HETvsWT.png", width = 8, height = 6, units = "in", res = 300)
MA_HETvsWT <-plotMA(res_HET_vs_WT, ylim = c(-3, 3), , main = "MA Plot: HET vs WT")
dev.off()
```

### PCA Plot

To construct the PCA plot, first we selected a normalization strategy. VST was chosen for its speed and efficiency, and the parameter blind = TRUE is better for quality control purposes as it detects unexpected patterns due to it not being biased toward experimental design. 

```{r normQC}
vsd <- vst(dds, blind=TRUE)
norm_count <- assay(vsd)
head(norm_count, 10)
```

```{r PCA}
png("PCA_plot.png", width = 8, height = 6, units = "in", res = 300)
pca_data <- plotPCA(vsd, intgroup="genotype", returnData=TRUE)
dev.off()
```
## Adding in gene names

Prior to this, we parsed through our GTF file to select the gene names and their IDs. Now, we will use it to construct DESeq2 tibbles with the gene names included. 
```{r id-name}
id2name <- as.tibble(read_tsv("ids_and_names.txt"))
head(id2name, 10)
```

KO vs WT:
```{r name-join_KO_WT}
res_KO_vs_WT  <- as.tibble(res_KO_vs_WT, rownames = 'gene') %>% left_join(id2name, by = c("gene" = "gene_id")) %>% relocate(gene_name)

head(res_KO_vs_WT, 10)
```


HET vs WT:
```{r name-join-view-HET_WT}
res_HET_vs_WT  <- as.tibble(res_HET_vs_WT, rownames = 'gene') %>% left_join(id2name, by = c("gene" = "gene_id")) %>% relocate(gene_name)

head(res_HET_vs_WT, 10)
```


## Enrichment Analysis: Performing GSEA (Broad Methodology)

Focusing on KO vs WT and HET vs WT for our GSEA analysis, ranked lists were generated by the Wald statistic.
```{r ranked}
ranked_list <- function(res) {
  ranks <- res$stat # Wald statistic (log2 fold change / standard error) 
  names(ranks) <- res$gene_name
  ranks <- ranks[!is.na(names(ranks)) & !is.na(ranks)] #remove NA
  ranks <- ranks[!duplicated(names(ranks))] #remove duplicates
  sort(ranks, decreasing = TRUE)
}

ranks_KO_vs_WT  <- ranked_list(res_KO_vs_WT)
ranks_HET_vs_WT <- ranked_list(res_HET_vs_WT)
```


To replicate the paper's methods, we use the MsigDB pathways for GO BP and GO CC. Genes of interest collected from the paper as well as its sources were used in place of GWAS HPO or DO. 
```{r load_gene_sets-GO:BP}
library(msigdbr)

msig_bp <- msigdbr(
  species = "Mus musculus",
  category = "C5",
  subcategory = "GO:BP"
)

pathways_bp <- split(msig_bp$gene_symbol, msig_bp$gs_name)
```

```{r load_gene_sets-GO:CC}
msig_cc <- msigdbr(
  species = "Mus musculus",
  category = "C5",
  subcategory = "GO:CC"
)
pathways_cc <- split(msig_cc$gene_symbol, msig_cc$gs_name)
```

```{r human2mouse}
orthologs <- read.delim("HOM_MouseHumanSequence.rpt.txt", header = TRUE, stringsAsFactors = FALSE)
```

```{r load_gene_sets-GWAS}

gwas_genesets <- list(
  "SCHIZOPHRENIA_GWAS" = c("DRD2", "CACNA1C", "TCF4", "ZNF804A", "NRGN", "GRM3", "GRIN2A", "CLCN3"),
  "BIPOLAR_GWAS" = c("CACNA1C", "ANK3", "ODZ4", "NCAN", "TRANK1", "PBRM1"),
  "AUTISM_GWAS" = c("CHD8", "SHANK3", "NLGN3", "NRXN1", "CNTNAP2", "SCN2A", "DYRK1A"),
  "DEPRESSION_GWAS" = c("SORCS3", "MEF2C", "NEGR1", "LRFN5", "ESR2"),
  "ADHD_GWAS" = c("DUSP6", "SEMA6D", "FOXP2", "ST3GAL3", "KDM4A"),
  "PARKINSONS_GWAS" = c("SNCA", "LRRK2", "GBA", "MAPT", "VPS35"),
  "ALZHEIMERS_GWAS" = c("APOE", "CLU", "PICALM", "BIN1", "CR1", "ABCA7"),
  "ALS_GWAS" = c("SOD1", "C9ORF72", "TARDBP", "FUS", "OPTN")
)

gwas_genesets_mouse <- lapply(gwas_genesets, function(genes) {
  mouse_genes <- orthologs$Symbol[orthologs$HGNC.symbol %in% genes]
  unique(mouse_genes) #convert human gene names to mouse gene names
})


# Check how many genes mapped per set
sapply(gwas_genesets_mouse, length)

```

### Running FGSEA: BP

Multilevel FGSEA was run on each set of pathways with comparisons KO vs WT and HET vs WT. A padj threshold of 0.25 was used, along with a minSize of 1 for GWAS to increase the range in order to hopefully visualize enrichment of these pathways in our data. 

KO vs WT
```{r fgseaBP-KO_WT}

# Run fgsea (multilevel)
fgseaResBP_KOvsWT <- fgsea(
  pathways = pathways_bp,
  stats    = ranks_KO_vs_WT,
  minSize  = 15,
  maxSize  = 500
)

# Filter significant pathways
fgseaResBP_KOvsWT_sig <- fgseaResBP_KOvsWT %>%
  filter(padj < 0.25)

# # Collapse redundant GO terms to remove redundancy 
collapsedBP_KOvsWT <- collapsePathways(
  fgseaResBP_KOvsWT_sig,
  pathways = pathways_bp,
  stats    = ranks_KO_vs_WT
)

fgseaResBP_KOvsWT_sig_collapsed <-
  fgseaResBP_KOvsWT_sig[
    pathway %in% collapsedBP_KOvsWT$mainPathways
  ]
```

HET vs WT
```{r fgseaBP-HET_WT}


fgseaResBP_HETvsWT <- fgsea(
  pathways = pathways_bp,
  stats    = ranks_HET_vs_WT,
  minSize  = 15,
  maxSize  = 500
)


fgseaResBP_HETvsWT_sig <- fgseaResBP_HETvsWT %>%
  filter(padj < 0.25)


collapsedBP_HETvsWT <- collapsePathways(
  fgseaResBP_HETvsWT_sig,
  pathways = pathways_bp,
  stats    = ranks_HET_vs_WT
)

fgseaResBP_HET_vs_WT_sig_collapsed <-
  fgseaResBP_HETvsWT_sig[
    pathway %in% collapsedBP_HETvsWT$mainPathways
  ]
```


### Running FGSEA: CC
KO vs WT
```{r fgseaCC-KO_WT}


fgseaResCC_KOvsWT <- fgsea(
  pathways = pathways_cc,
  stats    = ranks_KO_vs_WT,
  minSize  = 15,
  maxSize  = 500
)


fgseaResCC_KOvsWT_sig <- fgseaResCC_KOvsWT %>%
  filter(padj < 0.25)


collapsedCC_KOvsWT <- collapsePathways(
  fgseaResCC_KOvsWT_sig,
  pathways = pathways_cc,
  stats    = ranks_KO_vs_WT
)

fgseaResCC_KOvsWT_sig_collapsed <-
  fgseaResCC_KOvsWT_sig[
    pathway %in% collapsedCC_KOvsWT$mainPathways
  ]
```

HET vs WT
```{r fgseaCC-HET_WT}


fgseaResCC_HETvsWT <- fgsea(
  pathways = pathways_cc,
  stats    = ranks_HET_vs_WT,
  minSize  = 15,
  maxSize  = 500
)


fgseaResCC_HETvsWT_sig <- fgseaResCC_HETvsWT %>%
  filter(padj < 0.25)


collapsedCC_HETvsWT <- collapsePathways(
  fgseaResCC_HETvsWT_sig,
  pathways = pathways_cc,
  stats    = ranks_HET_vs_WT
)

fgseaResCC_HET_vs_WT_sig_collapsed <-
  fgseaResCC_HETvsWT_sig[
    pathway %in% collapsedCC_HETvsWT$mainPathways
  ]
```


### Running FGSEA: GWAS
KO vs WT
```{r fgseaGWAS-KO_WT}


fgseaResGWAS_KOvsWT <- fgsea(
  pathways = gwas_genesets,
  stats    = ranks_KO_vs_WT,
  minSize  = 1,
  maxSize  = 500
)


fgseaResGWAS_KOvsWT_sig <- fgseaResGWAS_KOvsWT %>%
  filter(padj < 0.25)


collapsedGWAS_KOvsWT <- collapsePathways(
  fgseaResGWAS_KOvsWT_sig,
  pathways = gwas_genesets,
  stats    = ranks_KO_vs_WT
)

fgseaResGWAS_KOvsWT_sig_collapsed <-
  fgseaResGWAS_KOvsWT_sig[
    pathway %in% collapsedGWAS_KOvsWT$mainPathways
  ]
```

HET vs WT
```{r fgseaGWAS-HET_WT}


fgseaResGWAS_HETvsWT <- fgsea(
  pathways = gwas_genesets,
  stats    = ranks_HET_vs_WT,
  minSize  = 1,
  maxSize  = 500
)


fgseaResGWAS_HETvsWT_sig <- fgseaResGWAS_HETvsWT %>%
  filter(padj < 0.25)


collapsedGWAS_HETvsWT <- collapsePathways(
  fgseaResGWAS_HETvsWT_sig,
  pathways = gwas_genesets,
  stats    = ranks_HET_vs_WT
)

fgseaResGWAS_HET_vs_WT_sig_collapsed <-
  fgseaResGWAS_HETvsWT_sig[
    pathway %in% collapsedGWAS_HETvsWT$mainPathways
  ]
```


## Replicating Figure 5c, Enrichment Results:

```{r top_fgseas}

#BP
top_pathwaysBP_KOvsWT <- bind_rows(
  fgseaResBP_KOvsWT_sig_collapsed %>% filter(NES > 0) %>% slice_max(NES, n=5),
  fgseaResBP_KOvsWT_sig_collapsed %>% filter(NES < 0) %>% slice_min(NES, n=5)
) %>% mutate(comparison = "KO vs WT")

top_pathwaysBP_HETvsWT <- bind_rows(
  fgseaResBP_HET_vs_WT_sig_collapsed %>% filter(NES > 0) %>% slice_max(NES, n=5),
  fgseaResBP_HET_vs_WT_sig_collapsed %>% filter(NES < 0) %>% slice_min(NES, n=5)
) %>% mutate(comparison = "HET vs WT")

#CC
top_pathwaysCC_KOvsWT <- bind_rows(
  fgseaResCC_KOvsWT_sig_collapsed %>% filter(NES > 0) %>% slice_max(NES, n=5),
  fgseaResCC_KOvsWT_sig_collapsed %>% filter(NES < 0) %>% slice_min(NES, n=5)
) %>% mutate(comparison = "KO vs WT")

top_pathwaysCC_HETvsWT <- bind_rows(
  fgseaResCC_HET_vs_WT_sig_collapsed %>% filter(NES > 0) %>% slice_max(NES, n=5),
  fgseaResCC_HET_vs_WT_sig_collapsed %>% filter(NES < 0) %>% slice_min(NES, n=5)
) %>% mutate(comparison = "HET vs WT")

#GWAS
top_pathwaysGWAS_KOvsWT <- bind_rows(
  fgseaResGWAS_KOvsWT_sig_collapsed %>% filter(NES > 0) %>% slice_max(NES, n=5),
  fgseaResGWAS_KOvsWT_sig_collapsed %>% filter(NES < 0) %>% slice_min(NES, n=5)
) %>% mutate(comparison = "KO vs WT")

top_pathwaysGWAS_HETvsWT <- bind_rows(
  fgseaResGWAS_HET_vs_WT_sig_collapsed %>% filter(NES > 0) %>% slice_max(NES, n=5),
  fgseaResGWAS_HET_vs_WT_sig_collapsed %>% filter(NES < 0) %>% slice_min(NES, n=5)
) %>% mutate(comparison = "HET vs WT")
```

```{r fgsea-plot}
all_top_pathways <- bind_rows(
  top_pathwaysBP_KOvsWT,
  top_pathwaysBP_HETvsWT,
  top_pathwaysCC_KOvsWT,
  top_pathwaysCC_HETvsWT,
  top_pathwaysGWAS_KOvsWT,
  top_pathwaysGWAS_HETvsWT
)

# Add category labels
all_top_pathways <- all_top_pathways %>%
  mutate(
    category = case_when(
      grepl("^GOCC_", pathway) ~ "GO Cellular Component",
      grepl("^GOBP_", pathway) ~ "GO Biological Process",
      grepl("_GWAS$", pathway) ~ "Human Disease GWAS",
      TRUE ~ "Other"
    ),
    pathway_clean = gsub("^GOCC_|^GOBP_|_GWAS$", "", pathway),
    pathway_clean = gsub("_", " ", pathway_clean)
  )

# Order categories
all_top_pathways$category <- factor(all_top_pathways$category, 
                                     levels = c("GO Cellular Component", 
                                               "GO Biological Process", 
                                               "Human Disease GWAS"))

# Plot
fgsea <- ggplot(all_top_pathways, aes(x = comparison, y = pathway_clean, 
                              size = -log10(padj), color = NES)) +
  geom_point() +
  scale_color_gradient2(low = "blue", mid = "white", high = "red", 
                        midpoint = 0, limits = c(-3, 3)) +
  scale_size_continuous(range = c(2, 10), name = "-Log10(FDR)") +
  facet_grid(category ~ ., scales = "free_y", space = "free_y") +
  theme_minimal() +
  labs(x = "", y = "", title = "FGSEA Enrichment: 12wk PFC") +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold", 
                             margin = margin(b = 20)),  # Added margin below title
    strip.text.y = element_text(angle = 0, hjust = 0, face = "bold", size = 10),
    axis.text.y = element_text(size = 8),
    axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
    panel.spacing = unit(1.5, "lines")  # More space between facets
  )

ggsave(
  filename = "fgsea_enrichment_plot.png",
  plot = last_plot(),
  width = 10,
  height = 12,  # Taller to accommodate all pathways
  dpi = 300,
  units = "in",
  bg = "white" 
)
```

## SessionInfo

This will print out a summary of the current packages installed in the environment.
```{r session-info}
sessioninfo::session_info()
```